patches/tModLoader/Terraria/Main.TML.cs

```diff
@@ -445,4 +449,50 @@ public static void AddSignalTraps()
 		PosixSignalRegistration.Create(PosixSignal.SIGINT, Handle);
 		PosixSignalRegistration.Create(PosixSignal.SIGTERM, Handle);
 	}
+
+	private static void ShowModsNeedReloadMenuInsteadIfNeeded()
+	{
+		// If any loaded mod or config differs from normal, intercept SinglePlayer menu with a prompt
+		// TODO: Should this remember the mod loadout before joining server? Right now it only cares about version downgrades. MP mods will still load.
+		bool needsReload = false;
+		List<ReloadRequiredExplanation> reloadRequiredExplanationEntries = new();
+		var normalModsToLoad = ModOrganizer.RecheckVersionsToLoad();
+		foreach (var loadedMod in ModLoader.ModLoader.Mods) {
+			if (loadedMod is ModLoaderMod || loadedMod.File == null)
+				continue;
+			var normalMod = normalModsToLoad.First(mod => mod.Name == loadedMod.Name); // If this throws, we have a big issue.
+			if (normalMod.modFile.path != loadedMod.File.path) {
+				reloadRequiredExplanationEntries.Add(new ReloadRequiredExplanation(1, normalMod.Name, normalMod, $"[c/FFFACD:Switch] to v{normalMod.Version}\n(Currently loaded v{loadedMod.Version})"));
+				needsReload = true;
+			}
+		}
+
+		if(ConfigManager.AnyModNeedsReloadCheckOnly(normalModsToLoad, reloadRequiredExplanationEntries)) {
+			// If reload is required, show message. Back action should leave current ModConfig instances unchanged 
+			needsReload = true;
+		}
+		else if (!needsReload) {
+			// Otherwise, config changes take effect immediately and the player select menu will be shown
+			ConfigManager.LoadAll(); // Makes sure MP configs are cleared.
+			ConfigManager.OnChangedAll();
+		}
```

Comment:
Better to separate this from the above `if`, since it's more logically distinct. Use an early `return`, change the method name to something which doesn't explicitly mention the reload menu. As is, it implies that nothing will be changed if the reload menu isn't done. Perhaps `PrepareLoadedModsAndConfigsForSingleplayer`

```suggestion

		if (!needsReload) {
			// Otherwise, config changes take effect immediately and the player select menu will be shown
			ConfigManager.LoadAll(); // Makes sure MP configs are cleared.
			ConfigManager.OnChangedAll();
			return;
		}
```